snowpack/src/plugins/plugin-optimize.ts

```diff
@@ -0,0 +1,131 @@
+import path from 'path';
+import glob from 'glob';
+import * as esbuild from 'esbuild';
+import {init, parse} from 'es-module-lexer';
+import {SnowpackConfig, SnowpackPlugin} from '../types/snowpack';
+import {appendHTMLToHead, getExt, relativeURL} from '../util';
+
+interface OptimizePluginOptions {
+  exclude?: string | string[];
+  minifyCSS?: boolean;
+  minifyHTML?: boolean;
+  minifyJS?: boolean;
+}
+
+/**
+ * Default optimizer for Snawpack, unless another one is given
+ */
+export function optimize(config: SnowpackConfig, options: OptimizePluginOptions): SnowpackPlugin {
+  async function optimizeFile({
+    esbuildService,
+    file,
+    modulesToPreload,
+    preloadModuleFile,
+  }: {
+    esbuildService: esbuild.Service;
+    file: string;
+    modulesToPreload: Set<string>;
+    preloadModuleFile: string;
+  }) {
+    const {baseExt} = getExt(file);
+
+    // optimize based on extension. if it’s not here, leave as-is
+    switch (baseExt) {
+      case '.css': {
+        // TODO: minify CSS
+        break;
+      }
+      case '.js':
+      case '.mjs': {
+        try {
+          let code = fs.readFileSync(file, 'utf-8');
+
+          // 1. take imports, add to preload
+          const [imports] = parse(code);
+          imports
+            .filter(({d}) => d === -1) // only preload static imports (this will be > -1 for dynamic)
+            .forEach(({s, e}) => {
+              modulesToPreload.add(path.resolve(path.dirname(file), code.substring(s, e)));
+            });
+
+          // 2. minify (if enabled)
+          if (options.minifyJS) {
+            const minified = await esbuildService.transform(code, {minify: true});
+            code = minified.js;
+            fs.writeFileSync(file, code);
+          }
+        } catch (err) {
+          throw new Error(`Trouble optimizing JS: ${file}\n${err.toString()}`);
+        }
+        break;
+      }
+      case '.html': {
+        // TODO: minify HTML
+        let code = fs.readFileSync(file, 'utf-8');
+
+        const preloadScript = `<link rel="modulepreload" href="${relativeURL(
+          path.dirname(file),
+          preloadModuleFile,
+        )}" />`;
+
+        code = appendHTMLToHead(code, preloadScript);
+        return fs.writeFileSync(file, code);
+        break;
+      }
+    }
+  }
+
+  function writeModulePreloadFile({
+    dest,
+    modulesToPreload,
+  }: {
+    dest: string;
+    modulesToPreload: Set<string>;
+  }) {
+    const sortedModules = [...modulesToPreload];
+    sortedModules.sort((a, b) => a.localeCompare(b));
+    const code = `import '${sortedModules
+      .map((url) => relativeURL(path.dirname(dest), url))
+      .join("';\nimport '")}';`;
+    fs.writeFileSync(dest, code, 'utf-8');
+  }
+
+  return {
+    name: '@snowpack/plugin-optimize',
+    async optimize({buildDirectory}) {
+      const esbuildService = await esbuild.startService();
+      await init;
+
+      // 1. scan directory
+      const allFiles = glob.sync('**/*', {
```

Comment:
Mirorring a concern you posted elsewhere: yea, this will definitely cause worse perf for a number of users. In the Skypack catalog project for instance, this would cause every JS file to load, regardless of which page you're on/visiting. In any SPA, this would cause every JS file to load as well.

Minifying files this way is fine, but for preload specifically, even for an MVP, we need to scan for and add preload tags to each HTML file seperately via something like:
- For every individual HTML file:
  - Scan its external JS scripts (v1.1: any inline JS imports as well)
  - Traverse through the tree of imports, collect all deep imports
  - Inject them all as modulepreload tags into the HTML page itself

The webpack plugin does something like this already, although it's lucky in that webpack takes care of the traversal for us.